packages/synapse-react-client/src/components/SchemaDrivenAnnotationEditor/field/AdditionalPropertiesSchemaField.tsx

```diff
@@ -0,0 +1,282 @@
+  const { SelectWidget } = registry.widgets
+
+  // The type determines which widget we show.
+  const [propertyType, setPropertyType] = useState(
+    guessPropertyType(convertToArray(formData)),
+  )
+
+  // If the property type is updated, store it in a new variable where we'll show a warning if data may be lost on coersion
+  const [nextPropertyType, setNextPropertyType] = useState(propertyType)
+
+  /**
+   * This effect is invoked whenever the user attempts to change the datatype of a custom annotation.
+   */
+  useEffect(() => {
+    function onNextPropertyTypeUpdate() {
+      if (Array.isArray(formData)) {
+        const dataIsEmpty =
+          formData.length === 0 ||
+          formData.every(item => item == null || item == '')
+        const coercedList = transformDataFromPropertyType(
+          formData,
+          nextPropertyType,
+        )
+        // if the data is empty or identical after conversion, then just update the property type
+        if (dataIsEmpty || nextPropertyType !== propertyType) {
+          if (isEqual(formData, coercedList)) {
+            setPropertyType(nextPropertyType)
+          }
+        }
+      }
+    }
+
+    onNextPropertyTypeUpdate()
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [nextPropertyType])
+
+  /**
+   * This effect is invoked whenever the propertyType changes.
+   */
+  useEffect(() => {
+    function coerceDataAndUpdateWidget() {
+      if (Array.isArray(formData)) {
+        if (formData.every(item => item == null)) {
+          onDropPropertyClick(name)(new CustomEvent('dropEmptyProperty'))
+        } else {
+          const coercedList = transformDataFromPropertyType(
+            formData,
+            nextPropertyType,
+          ) as unknown as T
+
+          // Data conversion is non-destructive or has been confirmed by the user
+          setPropertyType(nextPropertyType)
+          // Coerce the data to match the new type
+          onChange(coercedList)
+        }
+      }
+    }
+
+    coerceDataAndUpdateWidget()
+    // Don't add other properties to dependency array because we don't want to automatically coerce input
+    // i.e. Only coerce data when the type changes, which should only be on mount or when the user explicitly chooses a new type.
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [propertyType])
+
+  const itemsSchema = getSchemaForPropertyType(propertyType)
+
+  // If the additional property has not yet been coerced to an array, trigger an update to coerce and render nothing.
+  // We must do this because RJSF does not currently batch these kinds of updates
+  // See https://github.com/rjsf-team/react-jsonschema-form/issues/3367
+  if (!Array.isArray(formData)) {
+    onChange(convertToArray(formData) as unknown as T)
+    return <></>
+  }
+
+  return (
+    <AdditionalPropertyContextProvider
+      value={{
+        dropProperty: e => {
+          onDropPropertyClick(name)(e)
+        },
+      }}
+    >
```

Comment:
Added a custom context to pass this callback around, since changes in RJSF have made it so this callback is not accessible from where we need to invoke it.